flink-libraries/flink-table/src/main/scala/org/apache/flink/api/table/plan/nodes/dataset/DataSetCalc.scala

```diff
@@ -66,8 +67,28 @@ class DataSetCalc(
     super.explainTerms(pw).item("name", opName)
   }
 
+  override def computeSelfCost (planner: RelOptPlanner): RelOptCost = {
+
+    val child = this.getInput
+    val rowCnt = RelMetadataQuery.getRowCount(child)
+    val exprCnt = calcProgram.getExprCount
+    planner.getCostFactory.makeCost(rowCnt, rowCnt * exprCnt, 0)
```

Comment:
Calc is implemented as `MapFunction` and will never spill to disk or cause network traffic.